packages/core/mesh/teleport-extension-object-sync/src/object-sync.ts

```diff
@@ -30,15 +30,39 @@ export class ObjectSync {
     await this._ctx.dispose();
   }
 
-  download(id: string): Promise<DataObject> {
+  download({ id, ctx }: { id: string; ctx?: Context }): Promise<DataObject> {
     log('download', { id });
-    const trigger = entry(this._downloadRequests, id).orInsert(new Trigger<DataObject>()).value;
+
+    const existingRequest = this._downloadRequests.get(id);
+    if (existingRequest) {
+      existingRequest.counter++;
+      return existingRequest.trigger.wait();
+    }
+
+    const value = entry(this._downloadRequests, id).orInsert({
+      trigger: new Trigger<DataObject>(),
+      counter: 1,
+    }).value;
 
     for (const extension of this._extensions) {
       extension.updateWantListInASeparateTask(new Set(this._downloadRequests.keys()));
     }
 
-    return trigger.wait();
+    ctx?.onDispose(() => {
+      // Remove request if context is disposed and nobody else requests it.
+      const request = entry(this._downloadRequests, id).value;
```

Comment:
could be just this._downloadRequests.get(id) if youre not using other `entry` functions